extensions/amp-analytics/0.1/requests.js

```diff
@@ -0,0 +1,274 @@
+    this.baseUrl = dev().assert(request['baseUrl']);
+
+    /** @private {number} */
+    this.maxDelay_ = Number(request['maxDelay']) || 0;
+
+    /** @private {!./variables.VariableService} */
+    this.variableService_ = variableServiceFor(this.win);
+
+    /** @private {!../../../src/service/url-replacements-impl.UrlReplacements} */
+    this.urlReplacementService_ = Services.urlReplacementsForDoc(ampdoc);
+
+    /** @private {?Promise<string>} */
+    this.baseUrlPromise_ = null;
+
+    /** @private {?Promise<string>} */
+    this.baseUrlTemplatePromise_ = null;
+
+    /** @private {!Array<!Promise<string>>}*/
+    this.extraUrlParamsPromise_ = [];
+
+    /** @private {!../../../src/preconnect.Preconnect} */
+    this.preconnect_ = preconnect;
+
+    /** @private {!function(string, !JsonObject)} */
+    this.handler_ = handler;
+
+    /** @const @private {!Object|undefined} */
+    this.whiteList_ = isSandbox ? SANDBOX_AVAILABLE_VARS : undefined;
+
+    /** @private {?number} */
+    this.schedule_ = null;
+
+    /** @private {?JsonObject} */
+    this.lastTrigger_ = null;
+
+    /** @private {?Promise<string>} */
+    this.sendPromise_ = null;
+  }
+
+  /**
+   * Exposed method to send a request on event.
+   * Real ping may be batched and send out later.
+   * @param {?JsonObject} configParams
+   * @param {!JsonObject} trigger
+   * @param {!./variables.ExpansionOptions} expansionOption
+   * @return {!Promise<string>}
+   */
+  send(configParams, trigger, expansionOption) {
+    this.lastTrigger_ = trigger;
+    const triggerParams = trigger['extraUrlParams'];
+    if (!this.baseUrlPromise_) {
+      expansionOption.freezeVar('extraUrlParams');
+      this.baseUrlTemplatePromise_ =
+          this.variableService_.expandTemplate(this.baseUrl, expansionOption);
+      this.baseUrlPromise_ = this.baseUrlTemplatePromise_.then(baseUrl => {
+        return this.urlReplacementService_.expandAsync(
+            baseUrl, undefined, this.whiteList_);
+      });
+    };
+
+    this.extraUrlParamsPromise_.push(
+        this.expandExtraUrlParams_(configParams, triggerParams, expansionOption)
+            .then(expandExtraUrlParams => {
+              return this.urlReplacementService_.expandAsync(
+                  expandExtraUrlParams, undefined, this.whiteList_);
+            })
+            .then(finalExtraUrlParams => {
+              return finalExtraUrlParams;
+            }));
+
+    return this.trigger_();
+  }
+
+  /**
+   * Dispose function that clear reqeust handler state.
+   */
+  dispose() {
+    if (this.schedule_) {
+      this.win.clearTimeout(this.schedule_);
+    }
+    this.reset_();
+  }
+
+  /**
+   * Function that schedule the actual request send.
+   * @private
+   */
+  trigger_() {
+    if (this.maxDelay_ == 0) {
+      return this.fire_();
+    }
+    // If is batched
+    if (!this.schedule_) {
+      // schedule fire_ after certain time
+      return this.sendPromise_ = new Promise(resolve => {
```

Comment:
Batch timeouts are only set on send() -> trigger_() ? Looks like what it does now is:
1. Analytics event happens -> send
2. timeout set = maxDelay
3. timeout up -> fire
4. Clean up timeout (reset_)
5. Analytics event happens -> send

There's a time between 4 and 5 when the batch counter isn't running, but certain events like say timers are technically running continuously, so when they do fire and set that timeout the actual batch interval will be inflated by the that timer interval at max.

I think it's fine to wait until first request to set any timers, but we shouldn't clean everything up after firing. I'd at least use setInterval here instead so batching happens on a continuous interval.

Reply:
I think we are talking about the `intervalDecay` here. Yes, that's another feature I'll work on. I added a TODO here.